common/src/primitives/merkle/pos/mod.rs

```diff
@@ -0,0 +1,188 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::num::NonZeroUsize;
+
+use super::tree::tree_size::TreeSize;
+
+/// Given a binary tree with leaf-count as powers of 2, this struct represents a position in the tree.
+/// This also contains all the math required to convert position representations.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct NodePosition {
+    tree_size: TreeSize,
+    absolute_index: usize,
+}
+
+impl NodePosition {
+    pub fn from_abs_index(tree_size: TreeSize, absolute_index: usize) -> Option<Self> {
+        if (tree_size.get() + 1).count_ones() != 1 {
+            return None;
+        }
+
+        Some(Self {
+            tree_size,
+            absolute_index,
+        })
+    }
+
+    pub fn from_position(
+        tree_size: TreeSize,
+        level_from_bottom: usize,
+        index_in_level: usize,
+    ) -> Option<Self> {
+        let level_count = tree_size.level_count().get();
+        if level_from_bottom >= level_count {
+            return None;
+        }
+
+        // To help in seeing how these formulas were derived, see this table that represents values in the case tree.len() == 31 == 0b11111:
+        //  level     level_start  level_start in binary    index_in_level_size
+        //  0         0            00000                    16
+        //  1         16           10000                    8
+        //  2         24           11000                    4
+        //  3         28           11100                    2
+        //  4         30           11110                    1
+
+        let level_from_top = level_count - level_from_bottom;
+        // to get leading ones, we shift the tree size, right then left, by the level we need (see the table above)
+        let level_start = (tree_size.get() >> level_from_top) << level_from_top;
+        let absolute_index = level_start + index_in_level;
+        // max number of nodes in a level, in level_from_bottom
+        let index_in_level_size = if level_start > 0 {
+            1 << (level_start.trailing_zeros() - 1)
+        } else {
+            tree_size.leaf_count().get()
+        };
+        if index_in_level >= index_in_level_size {
+            return None;
+        }
+
+        Some(Self {
+            tree_size,
+            absolute_index,
+        })
+    }
+
+    pub fn tree_size(&self) -> TreeSize {
+        self.tree_size
+    }
+
+    pub fn abs_index(&self) -> usize {
+        self.absolute_index
+    }
+
+    pub fn position(&self) -> (usize, usize) {
```

Comment:
I'll write docs. You're right.

I'm too hesitant to create a separate type... have been like that since I created this. The object itself is supposed to convey this.